src/history/hs.c

```diff
@@ -1236,121 +1236,186 @@ __hs_delete_key_from_pos(
+             * return a different error instead.
              */
-            __wt_free(session, uri_itr);
-            if ((ret = __wt_metadata_btree_id_to_uri(session, btree_id, &uri_itr)) != 0)
-                WT_ERR_MSG(session, ret,
-                  "Unable to find btree-id %" PRIu32
-                  " in the metadata file for the associated history store key %s",
-                  btree_id,
-                  __wt_buf_set_printable(session, hs_key->data, hs_key->size, prev_hs_key));
-
-            WT_ERR(__wt_open_cursor(session, uri_itr, NULL, NULL, &data_cursor));
-            F_SET(data_cursor, WT_CURSOR_RAW_OK);
-        } else {
-            WT_ERR(__wt_compare(session, NULL, hs_key, prev_hs_key, &cmp));
-            if (cmp == 0)
-                continue;
-        }
-        WT_ERR(__wt_buf_set(session, prev_hs_key, hs_key->data, hs_key->size));
-        prev_btree_id = btree_id;
+            WT_ERR_MSG(session, WT_ERROR,
+              "the associated history store key %s was not found in the data store %s",
+              __wt_buf_set_printable(session, hs_key->data, hs_key->size, prev_hs_key),
+              session->dhandle->name);
+
+        /* Swap current/previous buffers. */
+        tmp = hs_key;
+        hs_key = prev_hs_key;
+        prev_hs_key = tmp;
+    }
+    WT_ERR_NOTFOUND_OK(ret, true);
+err:
+    __wt_scr_free(session, &hs_key);
+    __wt_scr_free(session, &prev_hs_key);
+    return (ret);
+}
 
-        /* Re-purpose the previous-key buffer on error, safe because we're about to error out. */
-        data_cursor->set_key(data_cursor, hs_key);
-        if ((ret = data_cursor->search(data_cursor)) == WT_NOTFOUND)
-            WT_ERR_MSG(session, ret,
-              "In %s, the associated history store key %s was not found in the data store", uri_itr,
-              __wt_buf_set_printable(session, hs_key->data, hs_key->size, prev_hs_key));
-        WT_ERR(ret);
+/*
+ * __wt_history_store_verify_one --
+ *     Verify the history store for the btree that is set up in this session. This must be called
+ *     when we are known to have exclusive access to the btree.
+ */
+int
+__wt_history_store_verify_one(WT_SESSION_IMPL *session)
+{
+    WT_CURSOR *cursor;
+    WT_CURSOR_BTREE cbt;
+    WT_DECL_RET;
+    WT_ITEM hs_key;
+    uint32_t btree_id;
+    int exact;
+
+    cursor = session->hs_cursor;
+    btree_id = S2BT(session)->id;
+
+    /*
+     * We are required to position the history store cursor. Set it to the first record of our btree
+     * in the history store.
+     */
+    memset(&hs_key, 0, sizeof(hs_key));
+    cursor->set_key(cursor, btree_id, &hs_key, 0, 0, 0, 0);
+    ret = cursor->search_near(cursor, &exact);
+    WT_ERR_NOTFOUND_OK(ret, true);
+    if (ret == 0 && exact < 0)
+        ret = cursor->next(cursor);
+
+    /* If we positioned the cursor there is something to verify. */
+    if (ret != WT_NOTFOUND) {
```

Comment:
I think this isn't safe, imagine `cursor.next` returned `WT_PANIC`.

Reply:
Fixed.